### PR TITLE
Add support for RFXCOM devices over TCP/IP

### PR DESCRIPTION
**Pull request by Jamstah**
This pull request adds everything required to use RFXCOM over TCP. I did some refactoring of the other connector classes to reduce code duplication, although there is still more that could be done.

Took some ideas from @ivanfmartinez pull request on openhab 1.9, so have added an author credit for him as well.

### CODE REVIEW

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package org.openhab.binding.rfxcom.internal.connector;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java]: could you please add the license header?

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.util.Arrays;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RFXComStreamReader extends Thread {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java]: please add Javadoc with author tag

**Comment by kaikreuzer** [comment]
lgtm in general, but I cannot test due to the lack of hardware.
@paulianttila Could you please review, test and comment as well?

**Comment by Jamstah** [comment]
Thanks for reviewing @kaikreuzer, have added license and javadoc to the commit.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComJD2XXConnector.java [hunk 3]
```diff
@@ -31,7 +28,7 @@
  *
  * @author Pauli Anttila - Initial contribution
  */
-public class RFXComJD2XXConnector implements RFXComConnectorInterface {
+public class RFXComJD2XXConnector extends RFXComBaseConnector {
 
     private static final Logger logger = LoggerFactory.getLogger(RFXComJD2XXConnector.class);
```
Comment (paulianttila) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComJD2XXConnector.java]: _listeners variable can be removed here

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import java.io.InterruptedIOException;
+import java.util.Arrays;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * RFXCOM connector for TCP/IP communication.
```
Comment (paulianttila) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java]: TCP/IP is misleading here as class is used also for serial communication

**Comment by paulianttila** [comment]
Thanks @Jamstah, good additions and improvements.

Changes looks good to me as well. Unfortunately, I can't test changes for a while, but if you @Jamstah tested also serial port communication, I'm fine to take changes in.

**Comment by Jamstah** [comment]
@paulianttila Thanks for the review, have made those changes.

I have tested over TCP/IP (using socat as documented in the commit) and with the manual serial connector. JD2XX doesn't work for me on my raspi so I haven't tested that. My test involved checking it would register, configure the RFXCOM, come online, and then receive messages. I haven't tested sending any messages because I don't need to do that in my configuration at the moment. Same for my other pull request.

I can potentially do more testing if I need to, but also, probably not for a while.

**Comment by kaikreuzer** [comment]
Thank you both!

**Comment by alco28** [comment]
It may be a stupid question, but How can I use this rfxcom tcpbridge? I'm running a rfxcom (433) tranceiver on a server with the Ser2Net service, and I can't find any documentation on how to set the bridge function in Openhab2 things or other configuration files... @paulianttila can you please make some examples or docs?

many thanks!

**Comment by Jamstah** [comment]
@alco28 how is your distribution installed? I don't think this has made it into an official release yet, so you'd need to build the binding and install that on your openhab2 server.

Once that's done, you can add it manually in the paper ui (RFXCOM USB Transceiver over TCP/IP). Its a rfxcom:tcpbridge.

**Comment by alco28** [comment]
Last distro from this Github, and I discovered that it's packaged in the latest release!
The configuration is a bit hidden in the PaperUI. You can't find it at the "brigde" tab / menu. but It's in the "things" part with at device called "RFXCOM USB Transceiver over TCP/IP". But no documentation about this way of configuration etc. So please add it to the wiki..

**Comment by Jamstah** [comment]
I haven't tried with ser2net, I used socat to make it available:
socat tcp-listen:10001,fork,reuseaddr file:/dev/ttyUSB0,raw

If you are configuring in the paper ui, you just need to put in the host you're running socat on and port number from the command line. I'm assuming ser2net would work in a similar way.

Which wiki were you expecting to be updated? If you can suggest UI updates that would make configuration easier, I'm sure that would be appreciated too, just raise an issue.

**Pull request merged**